ColumnFallback plugin to fallback to original column

shioyama opened the pull request:
Solves: #335

I wanted to see if this would work, and it does! This works like #335 but supports any backend. (Only lightly tested mind you.)

@mattboldt if you have time, have a look. I think with some tests and a bit of effort to work out the edge cases, this could be a really great new plugin. A lot of people want this.

shioyama commented on lib/mobility/plugins/original_column.rb:
```diff
@@ -0,0 +1,15 @@
+    end
+
+    register_plugin(:original_column, OriginalColumn)
+  end
```
I haven't really documented this anywhere yet, but this is a kind of convention/trick of defining a global plugin which does nothing, then defining a corresponding ORM-specific plugin which depends on the global one, and which is included automatically when you include the `active_record` plugin. So it's included but disabled by default.

That way, in your plugins, enabling the empty base plugin (`original_column`) indirectly enables the ORM-specific one. If in the future there was a corresponding plugin for Sequel, the same name could be used for that one (this is how `dirty`, `cache` and `query` work).

shioyama commented on lib/mobility/plugins/active_record/original_column.rb:
```diff
@@ -0,0 +1,60 @@
+
+        module BackendInstanceMethods
+          def read(locale, **)
+            if locale == I18n.default_locale
```
I don't love this dependence on `I18n` ... :thinking:

mattboldt replied:
I actually only used `I18n.default_locale` as an example. In practice, I actually need it to work with `en` OR `en-US`, where `en-US` falls back to plain `en`.

So this check would for sure be configurable -- either a static plugin config, or maybe a proc called in-context when building a new instance.

shioyama replied:
It could be optionally a proc which would be evaluated in context. The problem is that then we need to use `define_method` instead of `def` which has (small) perf implications, which I try to avoid on anything that runs through attribute getters/setters.

shioyama commented on lib/mobility/plugins/active_record/original_column.rb:
```diff
@@ -0,0 +1,60 @@
+          def build_node(attr, locale)
+            if locale == I18n.default_locale
+              # is the MobilityExpressions plugin necessary here?
+              model_class.arel_table[attr].extend(Plugins::Arel::MobilityExpressions)
```
Never got around to investigating this.

shioyama replied:
I don't _think_ this should be necessary.

shioyama commented on lib/mobility/plugins/original_column.rb:
```diff
@@ -0,0 +1,15 @@
+    module OriginalColumn
+      extend Plugin
+
+      default false
```
The default is `false` which means that it needs to be explicitly passed `original_column: true` on every call to `translates`. I think this makes sense since this plugin incurs a perf hit on every read/write, and probaly people only want to use it selectively on some columns/models.

mattboldt replied:
Oh! This may be the right spot to get rid of that `I18n` dependency.
```
class Post
  extend Mobility
  translates :name, type: :string, original_column_fallbacks: [:en, :en-US]
end
```

shioyama replied:
Yes indeed! I'll try that.

shioyama commented:
This is actually a really really useful plugin and super simple. Hoping we can get the details worked out and release in 1.2.

shioyama commented on lib/mobility/plugins/active_record/original_column.rb:
```diff
@@ -0,0 +1,60 @@
+
+          # def apply_scope
+          #   super
+          # end
```
This should be fine to remove; it will only add joins when it sees translated attribute nodes, but since `build_node` will use the column node for the default locale, it should not join for the default locale case.

shioyama commented:
Note that with the new API for querying with a block, this will work with stuff like:

```ruby
I18n.default_locale = :en
Post.i18n do
  title(:en).eq("foo").or(title(:ja).eq("bar"))
end.to_sql
```

```sql
SELECT "posts".* FROM "posts"
  LEFT OUTER JOIN "mobility_string_translations" "Post_title_ja_string_translations"
    ON "Post_title_ja_string_translations"."key" = 'test'
    AND "Post_title_ja_string_translations"."locale" = 'ja'
    AND "Post_title_ja_string_translations"."translatable_type" = 'Post'
    AND "Post_title_ja_string_translations"."translatable_id" = "posts"."id"
  WHERE ("posts"."title" = 'foo' OR "Post_title_ja_string_translations"."value" = 'bar')
```

Here the backend is KeyValue, but would work for any backend. Notice it's only joining for `ja` and not for `en` since `en` is the default locale.

shioyama commented on lib/mobility/plugins/active_record.rb:
```diff
@@ -4,6 +4,7 @@
 require_relative "./active_record/cache"
 require_relative "./active_record/query"
 require_relative "./active_record/uniqueness_validation"
+require_relative "./active_record/original_column"
```
We could maybe debate whether this is a good name.

mattboldt replied:
I was thinking the name could include the word "fallback" or "fallthrough" like with `fallthrough_accessors`, since the plugin only interferes if certain conditions are met. Maybe `fallthrough_original_column`, `original_column_fallthrough`?

shioyama replied:
Yeah hmm still thinking about it :thinking:

shioyama commented:
Thanks @stebo for the feedback! This is a great and easy win, so I'd like to give it a name and work out details, and get it out in a 1.2 release soon.

Klockner commented:
> Thanks @stebo for the feedback! This is a great and easy win, so I'd like to give it a name and work out details, and get it out in a 1.2 release soon.

Hello @shioyama, you still have plans to release this one? Thanks!

shioyama commented:
Yep! It's planned for the next release.

On June 30, 2021 3:26:30 AM GMT+09:00, Gabriel Klockner ***@***.***> wrote:
>> Thanks @stebo for the feedback! This is a great and easy win, so I'd
>like to give it a name and work out details, and get it out in a 1.2
>release soon.
>
>Hello @shioyama, you still have plans to release this one? Thanks!
>
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/shioyama/mobility/pull/512#issuecomment-870818653

shioyama commented:
So ok folks, sorry for the delay on this.

For the name, I'm thinking of just `column_fallback` (singular), so:

```ruby
class Post
  extend Mobility
  translates :name, type: :string, column_fallback: true # falls back on I18n.default_locale
  # translates :name, type: :string, column_fallback: [:en, :fr] # falls back if Mobility.locale is :en of :fr
  # translates :name, type: :string, column_fallback: ->(locale) { ... } # falls back if proc returns true
end
```

WDY(all)T?

peter-tran-goldenowl commented:
hi @shioyama . I have pull this branch and testing it but can not config it correctly. Could you please provide me a correct config? Here is my current config:

`translates :name,type: : string, original_column: true, original_column_fallbacks: [:en]`

gando001 commented:
Hi @shioyama is there an ETA on the next release including this PR?

shioyama commented:
Hi folks, haven't had time to get to this. Since it's a new thing (unlike bugs etc) I need to sit down and think about it a bit, so it's slightly downprioritized for me right now.

@peter-tran-goldenowl It's not working since it's not actually implemented yet :sweat_smile: I was just proposing an interface.

shioyama commented:
Hi folks! Sorry for the delay. I've rewritten this plugin supporting the boolean, locale array and proc options, and added tests. The way it is implemented each option type gets the best performance, so boolean just includes a module, locales array dynamically defines the `read` and `write` methods with `module_eval` (so dispatch is fast) and the proc option defines them with `define_method` (slowest, but unavoidable).

Tests are passing so once y'all try this out and are happy I should be able to release it quickly in a 1.2 release.

mattboldt commented:
@shioyama that's terrific! Thank you so much for your hard work on this. Gonna try to work this into our team's sprint and get it tested. If all goes well, it'll be in production in ~30 days! 🎉

shioyama commented:
I added a Sequel implementation, almost identical.

shioyama commented:
Oh just realized querying is not working for anything other than a `true` option value. I'll fix it and add specs covering that.

shioyama commented:
Ok added querying for all options in AR::ColumnFallback, with tests to check. Still need to finish other options for Sequel version.

benoror commented:
heads up/q: is this plugin documented somewhere?
